pkg/proxy/proxy.go

```diff
@@ -528,16 +528,21 @@ func (p *Proxy) SetProxyPort(name string, proxyType types.ProxyType, port uint16
 
 // ReinstallRoutingRules ensures the presence of routing rules and tables needed
 // to route packets to and from the L7 proxy.
-func (p *Proxy) ReinstallRoutingRules() error {
+func (p *Proxy) ReinstallRoutingRules(mtu int) error {
 	fromIngressProxy, fromEgressProxy := requireFromProxyRoutes()
 
+	// Use RouteMTU only when both ingress and egress proxy.
+	if !fromIngressProxy || !fromEgressProxy {
+		mtu = 0
+	}
```

Comment:
This reminds me: the reason in some situation we don't install 2005 rules (`!fromXgressProxy`), is tunnel routing mode already has `default dev cilium_host` to route proxy traffic. In this case we need to make sure those default routes rather than 2005 rules have correct MTU.

But it could be a bad idea to affect all the other normal from-host traffic, I need to think about this.

What did CI say? Any failure in tunnel scenarios for pod-to-pod-with-l7-policy-encryption?

Reply:
To be honest, I was not considering the overlay scenario when working on this patch. Trying to play locally with Overlay doesn't seem to report any error to connectivity even before this patch (ESP packet not fragmented). Though, I would like to replicate locally with the exact kernel that are failing and see what happens with this test enabled.

AFAIU, CI seems to confirm this, connectivity tests are passing. The plain text IPv6 packet we're recording should be an exception and should not indicate that the whole communication was plain text (right?): I would expect more than one entry in the bpftool output otherwise, let me know if I'm mistaken.

Reply:
> The plain text IPv6 packet we're recording should be an exception and should not indicate that the whole communication was plain text.

Yes. It's a transient issue coming from...maybe from pod deletion, policy creating and deleting....

Can you open another draft PR to run the pod2pod encryption test, but without any other changes you made here such as routing MTU. I suspect the leak isn't caused by your PR, it's been there since forever.

Reply:
Withdraw what I said above.... without your changes pod2pod encryption test can't pass connectivity checks....

Reply:
> Can you open another draft PR to run the pod2pod encryption test, but without any other changes you made here such as routing MTU. I suspect the leak isn't caused by your PR, it's been there since forever.

Yep sure, I wanted to test CI local to avoid opening a draft PR :joy: I'll open it in a minute. I suspect too it is not affected by our MTU patch, but still addressing it in this issue would be a good option.

Reply:
> Withdraw what I said above.... without your changes pod2pod encryption test can't pass connectivity checks....

Not in native routive, but it still should pass in geneve/vxlan (apparently not in vxlan), right?

Reply:
https://github.com/cilium/cilium/actions/runs/11273423172

1. `rhel8-20240404.144247` failing with same reason
2. `5.15-20240710.064909` failing but immediately after the 1st connectivity test (no after rotation)
3.  `bpf-next-20240711.013133` passing :open_mouth:

Reply:
Resolving for now, moving the whole conversation concerning the tests in a new issue+PR.